### PR TITLE
Fix update notifications for both global and local installation: Generate project hash and pass to checkpoint-client

### PR DESCRIPTION
**Pull request by 2color**
## What

- Calculate a SHA256 of the directory in which `schema.prisma` is found
- Calculate a SHA256 of the CLI file path

## Why

- Scope telemetry cache to CLI installation path (so update notifications work correctly for both global and local installation).
- Fix a bug where multiple projects reset the reminder duration and cause unexpected behaviour

## TODO

- [x] Bump checkpoint-client version once merged and released

### CODE REVIEW

File: src/packages/cli/src/bin.ts
Comment (Sytten) [inline src/packages/cli/src/bin.ts:234]: Left over comment?

File: src/packages/cli/src/bin.ts
Comment (matthewmueller) [inline src/packages/cli/src/bin.ts:212]: Is this the best spot for utilities @timsuchanek?

File: src/packages/cli/src/bin.ts
Comment (matthewmueller) [inline src/packages/cli/src/bin.ts:187]: Just to clarify, a `project` is an app, right? And you can have many projects that use the Prisma CLI `product`?

Comment (2color) [reply src/packages/cli/src/bin.ts:187]: Exactly

File: src/packages/cli/src/bin.ts
Comment (2color) [reply src/packages/cli/src/bin.ts:212]: This could be moved to `@prisma/sdk`

**Review by janpio — CHANGES_REQUESTED** [review]
Blocking the PR just to make sure it does not accidentally get merged before we have full agreement.

File: src/packages/cli/src/bin.ts
Comment (Jolg42) [inline src/packages/cli/src/bin.ts:191]: So one is named `cli_path_hash` the other `project`, why not `projectHash` or `project_path_hash`?

File: src/packages/cli/src/bin.ts
Comment (Jolg42) [inline src/packages/cli/src/bin.ts:217]: So the schema could be somewhere else because we have the arg `--schema=/path/to/myschema.prisma` so this could return null, then you do a hash of null?

File: src/packages/cli/src/bin.ts
Comment (Jolg42) [inline src/packages/cli/src/bin.ts:226]:
```suggestion
 * Get a unique identifier for the CLI installation path
```

File: src/packages/cli/src/bin.ts
Comment (Jolg42) [reply src/packages/cli/src/bin.ts:217]: Oh I was thinking you were using https://github.com/prisma/prisma/blob/902d4384fbadc9e27635c76e43f1bd4d25c0ba7e/src/packages/sdk/src/cli/getSchema.ts#L13 but you're using your own thing apparently

File: src/packages/cli/src/bin.ts
Comment (Jolg42) [inline src/packages/cli/src/bin.ts:229]: `return crypto...` returns a string not a Promise?

File: src/packages/cli/src/bin.ts
Comment (Jolg42) [inline src/packages/cli/src/bin.ts:241]: It returns a string not a Promise?

File: src/packages/cli/src/bin.ts
Comment (Jolg42) [inline src/packages/cli/src/bin.ts:216]: return crypto... returns a string not a Promise?

File: src/packages/cli/src/bin.ts
Comment (2color) [reply src/packages/cli/src/bin.ts:217]: Currently `getSchemaPath` will search for the `schema.prisma` file and default to the `cwd` if it can't find it.

As I read the comment I discovered that there's already a helper function which handles all cases: 
https://github.com/prisma/prisma/blob/c1624d9b953c8cf9396b3b7994f49a4fea5e9c95/src/packages/sdk/src/cli/getSchema.ts#L13 

Using it will require me to parse the `--schema` argument inside `bin.ts` and some commands don't have it. Since a project can exist prior to the schema existing (pre `prisma init` call) perhaps we should use the path of `package.json` as the path. 

Any thoughts?

Comment (Jolg42) [reply src/packages/cli/src/bin.ts:217]: The parsing is done here https://github.com/prisma/prisma/blob/c1624d9b953c8cf9396b3b7994f49a4fea5e9c95/src/packages/cli/src/bin.ts#L57

File: src/packages/cli/src/bin.ts
Comment (2color) [reply src/packages/cli/src/bin.ts:229]: That's a mistake. Will correct.

File: src/packages/cli/src/bin.ts
Comment (2color) [reply src/packages/cli/src/bin.ts:241]: Since it's async for the await calls, it does return a Promise.

File: src/packages/cli/src/bin.ts
Comment (Jolg42) [reply src/packages/cli/src/bin.ts:217]: @2color 
> Since a project can exist prior to the schema existing (pre prisma init call) perhaps we should use the path of package.json as the path.
The package.json may not exists as well actually

File: src/packages/cli/src/bin.ts
Comment (2color) [reply src/packages/cli/src/bin.ts:216]: The function is async as it calls `getSchemaPath`

File: src/packages/cli/src/bin.ts
Comment (2color) [reply src/packages/cli/src/bin.ts:191]: Renamed to `project_hash`

File: src/packages/cli/src/bin.ts
Comment (Jolg42) [reply src/packages/cli/src/bin.ts:216]: Maybe here https://github.com/prisma/prisma/blob/master/src/packages/sdk/src/cli/utils.ts

File: src/packages/cli/src/bin.ts
Comment (2color) [reply src/packages/cli/src/bin.ts:217]: I changed this to use the `getSchemaPath` method from the SDK and defaulting to `cwd` if it can't be found. This is fine from the checkpoint-client's perspective as its cache is based on the CLI installation path.

File: src/packages/sdk/src/cli/hashes.ts
Comment (Jolg42) [inline src/packages/sdk/src/cli/hashes.ts:12]: Because there is a throw here https://github.com/prisma/prisma/blob/902d4384fbadc9e27635c76e43f1bd4d25c0ba7e/src/packages/sdk/src/cli/getSchema.ts#L24-L27

That means `getProjectHash` won't work if `args['--schema']` is invalid.

In the end from the checkpoint perspective ignoring `args['--schema']` could be easier / better? @2color

Comment (Jolg42) [reply src/packages/sdk/src/cli/hashes.ts:12]: I guess the error would actually be thrown before reaching this anyway :D

**Review by Jolg42 — APPROVED** [review]

**Comment by Jolg42** [comment]
The CI errors here because this needs to be done `Bump checkpoint-client version once merged and released`

File: src/packages/cli/src/bin.ts
Comment (janpio) [inline src/packages/cli/src/bin.ts:188]: nit: also is a path hash, schema path to be precise

**Review by janpio — APPROVED** [review]
👍

File: src/packages/sdk/src/cli/hashes.ts
Comment (2color) [reply src/packages/sdk/src/cli/hashes.ts:12]: As discussed directly: 
If the --schema is invalid, I thought it won’t reach the checkpoint.check() call.

File: src/packages/cli/src/bin.ts
Comment (2color) [reply src/packages/cli/src/bin.ts:188]: That's true. 

I kept it that way so that consumers of checkpoint-client can decide what they use to identify a project. In the context of the checkpoint-client, I thought it'd be good to keep it projectHash as it could use a different value to identify a project. 

Nevertheless here it could be renamed to `schemaPathHash`

Comment (janpio) [reply src/packages/cli/src/bin.ts:188]: Good point, depending on context (client vs. server) it indeed can be differentiated.

**Review by timsuchanek — APPROVED** [review]

**Pull request merged**
